### PR TITLE
[Artifacts Tab] Rework External Artifact

### PR DESCRIPTION
**Pull request by fa9r**
## Describe changes

I reworked external artifacts to support fetching artifacts by version.

[Breaking Change] Since it's possible to fetch artifacts by name and version now, it no longer makes sense to additionally support the combination of model + model version + model artifact name + model artifact version, so I removed this option.

[Breaking Change] I also renamed the `artifact_name` argument to `name`.

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/stacks-and-components/component-guide) table and the [corresponding website section](https://zenml.io/integrations).
- [x] I have added tests to cover my changes.
- [ ] I have based my new branch on `develop` and the open PR is targeting `develop`. If your branch wasn't based on develop read [Contribution guide on rebasing branch to develop](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md#-pull-requests-rebase-your-branch-on-develop).

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: examples/e2e/pipelines/batch_inference.py [hunk 1]
```diff
@@ -59,16 +59,16 @@ def e2e_use_case_batch_inference():
...
-            artifact_name="preprocess_pipeline",
         ),
         target=target,
     )
     ########## DataQuality stage  ##########
     report, _ = evidently_report_step(
         reference_dataset=ExternalArtifact(
+            name="dataset_trn",
             pipeline_name=MetaConfig.pipeline_name_training,
-            artifact_name="dataset_trn",
```
Comment (avishniakov) [inline examples/e2e/pipelines/batch_inference.py:71]: These changes are not effective, cause E2E template is auto-updated once tests pass and to make them pass we need a fix on template side + new release tag here.
That's why test is failing.

File: examples/e2e/run.py [hunk 2]
```diff
@@ -190,8 +190,8 @@ def main(
     )
 
     artifact = ExternalArtifact(
+        name="predictions",
         pipeline_name=MetaConfig.pipeline_name_batch_inference,
-        artifact_name="predictions",
```
Comment (avishniakov) [inline examples/e2e/run.py:194]: same

**Review by avishniakov — CHANGES_REQUESTED** [review]
There is conceptual misunderstanding for me here. Can you explain with a code sample how I could achieve the same using new interface?
```python
from zenml import step, pipeline, get_step_context
from zenml.model import ModelConfig
from zenml.artifacts.external_artifact import ExternalArtifact

@step
def producer_step()->int:
    return 1

@step
def consumer_step(a:int):
    print(a)

@pipeline(model_config=ModelConfig(name="foo",create_new_model_version=True))
def producer():
    producer_step()

@pipeline(model_config=ModelConfig(name="foo"))
def consumer():
    consumer_step(ExternalArtifact(model_artifact_name="output"))

if __name__=="__main__":
    producer()
    consumer()
```

File: src/zenml/artifacts/external_artifact.py [hunk 3]
```diff
@@ -95,40 +98,19 @@ def my_pipeline():
...
+            values.get(field, None) is not None
+            for field in ["value", "id", "name"]
+        ]
+        if sum(options) > 1:
             raise ValueError(
-                "Only a value, an ID, pipeline/artifact name pair or "
-                "model name/model version/model artifact name group can be "
-                "provided when creating an external artifact."
+                "Only a value, an ID, or an artifact name can be provided "
+                "when creating an external artifact."
```
[the remark this answers is not in the document]
Comment (fa9r) [reply src/zenml/artifacts/external_artifact.py:108]: Yes, isn't it semantically the same? What would the current wording mean to you?

File: examples/e2e/pipelines/batch_inference.py (hunk 1 above)
Comment (fa9r) [reply examples/e2e/pipelines/batch_inference.py:71]: Ok then I undo the changes in `examples/e2e` and change them in the e2e repo instead?

**Comment by fa9r** [comment]
@avishniakov 

> Can you explain with a code sample how I could achieve the same using the new interface?

```python
@pipeline
def producer():
    producer_step()

@pipeline
def consumer():
    consumer_step(ExternalArtifact(name="producer::producer_step::output"))

if __name__=="__main__":
    producer()
    consumer()
```

**Comment by fa9r** [comment]
Discussion: add back (model + model version + artifact name) option

File: src/zenml/artifacts/external_artifact.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (fa9r) [reply src/zenml/artifacts/external_artifact.py]: Anyways, adjust it to your wording

File: examples/e2e/pipelines/batch_inference.py (hunk 1 above)
Comment (avishniakov) [reply examples/e2e/pipelines/batch_inference.py:71]: Yes, pelase

**Comment by fa9r** [comment]
@avishniakov I undid the e2e example changes. However, I haven't adjusted that template yet since I will do that later once we have the `model_version.get_artifact(artifact_name)`

**Pull request merged**